test/test_train_mnist_tensor.py

```diff
@@ -0,0 +1,157 @@
+import shutil
+import time
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torchvision import datasets, transforms
+import torch_xla
+import torch_xla_py.utils as xu
+import torch_xla_py.xla_model as xm
+import unittest
+
+
+class MNIST(nn.Module):
+
+  def __init__(self):
+    super(MNIST, self).__init__()
+    self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
+    self.bn1 = nn.BatchNorm2d(10)
+    self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
+    self.bn2 = nn.BatchNorm2d(20)
+    self.fc1 = nn.Linear(320, 50)
+    self.fc2 = nn.Linear(50, 10)
+
+  def forward(self, x):
+    x = F.relu(F.max_pool2d(self.conv1(x), 2))
+    x = self.bn1(x)
+    x = F.relu(F.max_pool2d(self.conv2(x), 2))
+    x = self.bn2(x)
+    x = x.view(-1, 320)
+    x = F.relu(self.fc1(x))
+    x = self.fc2(x)
+    return F.log_softmax(x, dim=1)
+
+
+def train_mnist():
+  assert FLAGS.num_cores == 1
+  torch.manual_seed(1)
+  # Training settings
+  lr = 0.01
+  momentum = 0.5
+  log_interval = 5
+
+  if FLAGS.fake_data:
+    train_loader = xu.SampleGenerator(
+        data=torch.zeros(FLAGS.batch_size, 1, 28, 28),
+        target=torch.zeros(FLAGS.batch_size, dtype=torch.int64),
+        sample_count=60000 // FLAGS.batch_size)
+    test_loader = xu.SampleGenerator(
+        data=torch.zeros(FLAGS.batch_size, 1, 28, 28),
+        target=torch.zeros(FLAGS.batch_size, dtype=torch.int64),
+        sample_count=10000 // FLAGS.batch_size)
+  else:
+    train_loader = torch.utils.data.DataLoader(
+        datasets.MNIST(
+            FLAGS.datadir,
+            train=True,
+            download=True,
+            transform=transforms.Compose([
+                transforms.ToTensor(),
+                transforms.Normalize((0.1307,), (0.3081,))
+            ])),
+        batch_size=FLAGS.batch_size,
+        shuffle=True,
+        num_workers=FLAGS.num_workers)
+    test_loader = torch.utils.data.DataLoader(
+        datasets.MNIST(
+            FLAGS.datadir,
+            train=False,
+            transform=transforms.Compose([
+                transforms.ToTensor(),
+                transforms.Normalize((0.1307,), (0.3081,))
+            ])),
+        batch_size=FLAGS.batch_size,
+        shuffle=True,
+        num_workers=FLAGS.num_workers)
+
+  device = xm.xla_device()
+  model = MNIST().to(device=device)
+  optimizer = optim.SGD(model.parameters(), lr=lr, momentum=momentum)
+  loss_fn = nn.NLLLoss()
+  accuracy = None
+  for epoch in range(1, FLAGS.num_epochs + 1):
+    # Training loop for epoch.
+    start_time = time.time()
+    processed = 0
+    for batch_idx, (data, target) in enumerate(train_loader):
+      if data.size()[0] != FLAGS.batch_size:
+        break
+      data = data.to(device=device)
+      target = target.to(device=device)
+
+      optimizer.zero_grad()
+      y = model(data)
+      loss = loss_fn(y, target)
+      loss.backward()
+      xm.optimizer_step(optimizer)
```

Comment:
What do you think about making this look more idiomatic? Instead of calling the `optimizer_step` like this, have a wrapper class with a step method. No need to address this right now, just curious whether there's a fundamental reason you went for this instead.

Reply:
Sure, that is another way. Though, if the user wants to interact with the optimizer core object (calling other APIs, accessing its fields), his code needs to be changed in many places.

(comment in wrong place)